nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/ui/provenance-event-listing/provenance-event-table/lineage/lineage.component.ts

```diff
@@ -0,0 +1,975 @@
+                    return !selection.empty();
+                },
+                clazz: 'fa fa-long-arrow-right',
+                text: 'Go to component',
+                action: function (selection: any, self: LineageComponent) {
+                    const selectionData: any = selection.datum();
+                    self.goToProvenanceEventSource.next({
+                        eventId: selectionData.id
+                    });
+                }
+            },
+            {
+                condition: function (selection: any, self: LineageComponent) {
+                    if (selection.empty()) {
+                        return false;
+                    }
+
+                    const selectionData: any = selection.datum();
+                    return self.supportsExpandCollapse(selectionData);
+                },
+                clazz: 'fa fa-binoculars',
+                text: 'Find parents',
+                action: function (selection: any, self: LineageComponent) {
+                    const selectionData: any = selection.datum();
+
+                    // TODO - cluster node id
+                    self.submitLineageQuery.next({
+                        lineageRequestType: 'PARENTS',
+                        eventId: selectionData.id
+                        // clusterNodeId: clusterNodeId
+                    });
+                }
+            },
+            {
+                condition: function (selection: any, self: LineageComponent) {
+                    if (selection.empty()) {
+                        return false;
+                    }
+
+                    const selectionData: any = selection.datum();
+                    return self.supportsExpandCollapse(selectionData);
+                },
+                clazz: 'fa fa-plus-square',
+                text: 'Expand',
+                action: function (selection: any, self: LineageComponent) {
+                    const selectionData: any = selection.datum();
+
+                    // TODO - cluster node id
+                    self.submitLineageQuery.next({
+                        lineageRequestType: 'CHILDREN',
+                        eventId: selectionData.id
+                        // clusterNodeId: clusterNodeId
+                    });
+                }
+            },
+            {
+                condition: function (selection: any, self: LineageComponent) {
+                    if (selection.empty()) {
+                        return false;
+                    }
+
+                    const selectionData: any = selection.datum();
+                    return self.supportsExpandCollapse(selectionData);
+                },
+                clazz: 'fa fa-minus-square',
+                text: 'Collapse',
+                action: function (selection: any, self: LineageComponent) {
+                    const selectionData: any = selection.datum();
+                    self.collapseLineage(selectionData);
+                }
+            }
+        ]
```

Comment:
I would try to simplify by not requiring the `self` references. You can accomplish this by using arrow functions, then `this` is the LineageComponent.
```suggestion
        menuItems: [
            {
                condition: (selection: any) => selection.empty(),
                clazz: 'fa fa-long-arrow-left',
                text: 'Back to events',
                action: (selection: any) => {
                    this.closeLineage.next();
                }
            },
            {
                condition: (selection: any) => !selection.empty(),
                clazz: 'fa fa-info-circle',
                text: 'View details',
                action: (selection: any) => {
                    const selectionData: any = selection.datum();

                    // TODO cluster node id
                    this.openEventDialog.next({
                        id: selectionData.id
                    });
                }
            },
            {
                condition: (selection: any) => !selection.empty(),
                clazz: 'fa fa-long-arrow-right',
                text: 'Go to component',
                action: (selection: any) =>  {
                    const selectionData: any = selection.datum();
                    this.goToProvenanceEventSource.next({
                        eventId: selectionData.id
                    });
                }
            },
            {
                condition: (selection: any) => {
                    if (selection.empty()) {
                        return false;
                    }

                    const selectionData: any = selection.datum();
                    return this.supportsExpandCollapse(selectionData);
                },
                clazz: 'fa fa-binoculars',
                text: 'Find parents',
                action: (selection: any) => {
                    const selectionData: any = selection.datum();

                    // TODO - cluster node id
                    this.submitLineageQuery.next({
                        lineageRequestType: 'PARENTS',
                        eventId: selectionData.id
                        // clusterNodeId: clusterNodeId
                    });
                }
            },
            {
                condition: (selection: any) => {
                    if (selection.empty()) {
                        return false;
                    }

                    const selectionData: any = selection.datum();
                    return this.supportsExpandCollapse(selectionData);
                },
                clazz: 'fa fa-plus-square',
                text: 'Expand',
                action: (selection: any) => {
                    const selectionData: any = selection.datum();

                    // TODO - cluster node id
                    this.submitLineageQuery.next({
                        lineageRequestType: 'CHILDREN',
                        eventId: selectionData.id
                        // clusterNodeId: clusterNodeId
                    });
                }
            },
            {
                condition: (selection: any) => {
                    if (selection.empty()) {
                        return false;
                    }

                    const selectionData: any = selection.datum();
                    return this.supportsExpandCollapse(selectionData);
                },
                clazz: 'fa fa-minus-square',
                text: 'Collapse',
                action: (selection: any) => {
                    const selectionData: any = selection.datum();
                    this.collapseLineage(selectionData);
                }
            }
        ]
```